enhancements/storage/csi-driver-install.md

```diff
@@ -0,0 +1,149 @@
+1. It is simple to create a opt-in installable driver operator.
+2. Providing an optional UI for configuration via OLM is a plus.
+3. Each driver's operator sits in its own repository and managing releases could be simpler.
+
+Cons:
+1. When these CSI drivers have to be installed by default, this approach causes a coupling between CVO managed cluster-storage-operator and driver's operator.
+2. In general there are some concerns about OLM's lack of documentation and everyone's understanding of it.
+3. Not sure about disconected cluster installation / update.
+
+### Installation via CVO
+
+We are also considering if these 5 CSI drivers should be installed by default and managed via CVO. We are considering using cluster-storage-operator to detect
+cloudprovider on which cluster is running and install necessary CSI driver by default. This does mean that, CSI driver is always installed along-side in-tree drivers.
+
+It makes it simpler to enable these drivers by default when in-tree to CSI migration goes GA.
+
+Expected workflow:
+1. CVO installs cluster-storage-operator.
+2. cluster-storage-operator checks on which cloud it is, let's say it's AWS.
+3. cluster-storage-operator starts AWS EBS CSI driver operator and sets its cluster-storage-operator status to Available + Progressing. In future it could run also AWS EFS CSI driver operator.
+4. AWS EBS CSI driver operator installs EBS CSI driver and creates some status CR (clusteroperators.config.openshift.io again?) that it's Available + Progressing.
+5. AWS EBS CSI driver operator checks that the driver is fully installed, i.e. driver DaemonSet and Deployment have full set of replicas and sets its own status to Available.
+6. cluster-storage-operator monitors status of AWS EBS CSI driver operator and sets its own status to Available.
+
+User cannot uninstall the driver, therefore the AWS EBS CSI driver operator does not consume any CR (whose deletion would tell the operator to uninstall the driver).
+
+Pros:
+1. Simpler handling of in-tree to CSI migration.
+2. Drivers are automatically installed and admin can monitor progress.
+
+Cons:
+1. If the CSI driver fails to deploy by the cluster-storage-operator, it will make the cluster-storage-operator degraded and as a result cluster will be degraded. This may not be optimal because initially we want CSI driver to be optional component.
+2. Supporting future configuration of drivers is tricky. Currently cluster-storage-operator is not a configurable operator. But enabling and disabling
+migration and making driver configurable will require designing CRDs which could support these functions.
+3. cluster-storage-operator needs to give RBAC permissions to the CSI operators. I.e. it must have all the permissions already. They may be quite powerful (privileged pods, edit PVs/PVCs) and may be slightly different for each CSI driver. Collection of these RBAC rules may be complicated, as they may be in multiple repos.
+
+### Installation with cloud providers
```

Comment:
I don't understand if there is a 1-1 mapping of cloud-provider to CSI.  Openstack seems to have 2 CSI plugins.  Then there's the possibility of things like Ceph?  It's not clear that all CSIs that we support will be tightly coupled to a specific cloud.  Also, we need to account for BYO host installs.  Someone might be on VMWare but not using the vsphere cloud provider for one reason or another, but might want the corresponding CSI driver.

I need to know more about the CSI story overall before I can say yes or no to coupling here.

Reply:
Right now, *every* in-tree cloud provider has at least one related volume plugin. IMO we can safely assume that every external cloud will need one or more related CSI drivers installed on the same cluster.

It does not work the same the other way around, some CSI drivers can (and will) run on any clusters (incl. bare metal). E.g. Gluster or Ceph, or anything based on NFS / iSCSI / FibreChannel.

Reply:
> Someone might be on VMWare but not using the vsphere cloud provider for one reason or another

I'd like to provide the same experience as with in-tree volume plugins. VMware volume plugin is always set up on vSphere cluster. Similarly, VMware CSI driver(s) will be installed. You don't need to use them.